### PR TITLE
chore: add E2E tests to puppeteer runner

### PR DESCRIPTION
**Pull request by dpgraham**
Add workflow that checks out `saucectl` and uses it to run sample config.yml's against this docker container

**WHY** This makes it so that we can't commit changes to the docker container without running tests that a) build the container; b) runs saucectl successful using this container

_ADDITIONS_
* `test/fixtures` that contains sample Cypress tests
  * `test/fixtures/basic/` is a basic test
  * `test/fixtures/typescript/` has a tsconfig.json and a ts file that get transpiled
* `test/e2e/*-config.yml`
  * contains sauce config ymls that target the `test/fixtures/`
* `test/e2e/run.sh`
  * runs `saucectl run --config ...` on all of the configs in `test/e2e`
  * transpiles the Typescript example

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -37,6 +40,8 @@ COPY --chown=seluser:seluser . .
 # Workaround for permissions in CI if run with a different user
 RUN chmod 777 -R /home/seluser/
 
+RUN ls -lah ./
```
Comment (alexplischke) [inline Dockerfile:43]: Debug leftover?

File: Dockerfile [hunk 2]
```diff
@@ -16,8 +16,11 @@ ENV PATH="/home/seluser/bin:/home/seluser/.nvm/versions/node/v${NODE_VERSION}/bi
 
 WORKDIR /home/seluser
 
+USER seluser
```
Comment (alexplischke) [inline Dockerfile:19]: Not necessary, as the base image is running seluser.

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+// For format details, see https://aka.ms/vscode-remote/devcontainer.json or this file's README at:
```
Comment (alexplischke) [inline .devcontainer/devcontainer.json:1]: Do we really want to include IDE specific files now?

File: .config/jest.config.js [hunk 4]
```diff
@@ -12,8 +12,8 @@ const readFile = promisify(fs.readFile)
 const DefaultRunCfg = {
     projectPath: `${HOME_DIR}`,
     match: [
-        `${HOME_DIR}/tests/?(*.)+(spec|test).[jt]s?(x)`,
-        `${HOME_DIR}/tests/**/?(*.)+(spec|test).[jt]s?(x)`
+        `${HOME_DIR}/tests/?(*.)+(spec|test).js?(x)`,
```
Comment (alexplischke) [inline .config/jest.config.js:15]: I guess this is something we should adjust for all images then?

Comment (dpgraham) [reply .config/jest.config.js:15]: @alexplischke not the TestCafe image. TestCafe accepts .ts as-is

File: .devcontainer/devcontainer.json (hunk 3 above)
Comment (dpgraham) [reply .devcontainer/devcontainer.json]: @alexplischke my bad, I took it out.

File: .config/jest.config.js (hunk 4 above)
Comment (christian-bromann) [reply .config/jest.config.js:16]: @dpgraham do we not expect any `.ts` files?

File: tests/typescript/tests/example.spec.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+describe('saucectl demo test', () => {
+    test('verify ts is transpiling', async function () {
+        // Just transpile a typescript variable to confirm that it's transpiling typescript
+        const str:String = "Hello World";
```
Comment (christian-bromann) [inline tests/typescript/tests/example.spec.ts:4]: Strings are usually defined with lowercase ([#ref](https://www.typescriptlang.org/docs/handbook/basic-types.html#string))

```suggestion
        const str:string = "Hello World";
```

**Review by vrunoa — APPROVED** [review]

File: .config/jest.config.js (hunk 4 above)
Comment (dpgraham) [reply .config/jest.config.js:16]: @christian-bromann you can't run .ts files directly, they need to be transpiled first.

File: .config/jest.config.js [hunk 6]
```diff
@@ -26,21 +27,40 @@ async function loadRunConfig(cfgPath) {
...
 
-function resolveTestMatches(runCfg) {
+function resolveTestMatches(HOME_DIR, runCfg) {
     return runCfg.match.map(
         p => {
             if (path.isAbsolute(p)) {
                 return p
             }
-            return path.join(runCfg.projectPath, p)
+            return path.join(HOME_DIR, runCfg.projectPath, p)
```
Comment (alexplischke) [inline .config/jest.config.js:36]: The project path is absolute. No need to join with the home directory.

**Review by alexplischke — CHANGES_REQUESTED** [review]

File: .config/jest.config.js (hunk 6 above)
Comment (dpgraham) [reply .config/jest.config.js]: @alexplischke I fixed it ... I noticed that the `run.yaml` 's in Cypress allow relative paths for `projectPath` (see https://github.com/saucelabs/sauce-cypress-runner/blob/master/tests/config-tests/run.yaml). So I just added a check that if it isn't absolute, make it absolute.

**Comment by dpgraham** [comment]
Added some fixes, should be ready but can't fully test it because of the permissions issue. May need to wait on saucectl before updating this.

**Review by alexplischke — APPROVED** [review]

**Pull request merged**